app/command_autocomplete.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"strings"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+// GetSuggestions returns suggestions for user input.
+func (a *App) GetSuggestions(commands []*model.AutocompleteData, userInput, roleID string) []model.AutocompleteSuggestion {
+	suggestions := []model.AutocompleteSuggestion{}
+	index := strings.Index(userInput, " ")
+	if index == -1 { // no space in user input
+		for _, command := range commands {
+			if strings.HasPrefix(command.Trigger, userInput) && (command.RoleID == roleID || roleID == model.SYSTEM_ADMIN_ROLE_ID || roleID == "") {
+				suggestions = append(suggestions, model.AutocompleteSuggestion{Hint: command.Trigger, Description: command.HelpText})
+			}
+		}
+		return suggestions
+	}
+	for _, command := range commands {
+		input := userInput[index+1:]
+		if command.Trigger != userInput[:index] {
+			continue
+		}
+		if roleID != "" && roleID != model.SYSTEM_ADMIN_ROLE_ID && roleID != command.RoleID {
+			continue
+		}
+		if len(command.Arguments) == 0 {
+			// Seek recursively in subcommands
+			subSuggestions := a.GetSuggestions(command.SubCommands, userInput[index+1:], roleID)
+			suggestions = append(suggestions, subSuggestions...)
+			continue
+		}
+		if command.Arguments[0].Name == "" { //Positional argument
+			for _, arg := range command.Arguments {
+				if arg.Type == model.AutocompleteTextArgType {
+					found, changedInput, suggestion := parseInputTextArgument(arg, input)
+					if found {
+						suggestions = append(suggestions, suggestion)
+						break
+					}
+					input = changedInput
+				} else if arg.Type == model.AutocompleteStaticListArgType {
+					found, changedInput, StaticListsuggestions := parseStaticListArgument(arg, input)
+					if found {
+						suggestions = append(suggestions, StaticListsuggestions...)
+						break
+					}
+					input = changedInput
+				} else if arg.Type == model.AutocompleteDynamicListArgType {
+					// TODO https://mattermost.atlassian.net/browse/MM-21491
+				}
+			}
+		} else { // Named argument
+			//TODO https://mattermost.atlassian.net/browse/MM-23194
+		}
+	}
+	return suggestions
+}
+
+func parseInputTextArgument(arg *model.AutocompleteArg, userInput string) (bool, string, model.AutocompleteSuggestion) {
+	in := strings.TrimPrefix(userInput, " ")
+	a := arg.Data.(*model.AutocompleteTextArg)
+	if len(in) == 0 { //typing of the argument is not started
```

Comment:
0/5 `The user has not started typing the argument.` This can be applied to other similar comments as well.